### ISSUE TITLE
How to register ScalarTypes?

### ISSUE THREAD

**Issue by brunobg** [webonyx/graphql-php#681]
How do I resolve extended ScalarTypes? https://webonyx.github.io/graphql-php/type-system/scalar-types/ explains how to create a class for them and the tests show how to ]build a Schema manually with them https://github.com/webonyx/graphql-php/blob/1f0ec2408010e01879b1f980afac65de2344c631/tests/Executor/VariablesTest.php#L132, but I can't find how to load them with a parser. 

Concrete example, trying to parse this GraphQL:

```graphql
scalar TestScalarType @scalar(class: "ModelariumTests\\TestScalarType")

schema {
  query: TestType
}

input TestInputObject {
  t: TestScalarType
}

type TestType {
  fieldWithObjectInput(input: TestInputObject): String
  fieldWithScalarInput(input: TestScalarType): TestScalarType
}
```

The idea is to use the directive to resolve the class. I'm parsing it like this:

```php
class TestScalarType extends ScalarType { ... }

class MyTest {
    public function testObjectParserQuery()
    {
        $document = Parser::parse(
            'query q($input: TestInputObject) {
            fieldWithObjectInput(input: $input)
        }'
        );

        $data = file_get_contents('test.graphql'); // loads the schema above
        $ast = \GraphQL\Language\Parser::parse($data);
        $schemaBuilder = new \GraphQL\Utils\BuildSchema(
            $ast,
            [__CLASS__, 'extendDatatypes']
        );
        $schema = $schemaBuilder->buildSchema();

        $result = Executor::execute($schema, $document, null, null, ['input' => [ 't' => 'valid']]);

        $this->assertEquals([], $result->errors);
        $expected = [
            'data' => ['fieldWithObjectInput' => '{"t":"xvalid"}'],
        ];
        $this->assertEquals($expected, $result->toArray());
    }
```

`extendDatatypes` is called for `TestType`, but I don't see how to load the scalar classes dynamically. Any hints please?

**Comment by spawnia** [comment]
What you are trying to do is quite tricky and requires deeply modifying the schema build process. Check out https://github.com/nuwave/lighthouse/tree/master/src/Schema for an example of how this can be achieved.

This library contains the necessary primitives to handle directives and customize schema building, and i do not think it should do anything more opinionated.

**Comment by brunobg** [comment]
Since you have a type loader for `buildschema`, I was expecting the lib would do something for any node kind or a pointer of how to do it. For other people that one day may find themselves here, I solved it by parsing the AST normally, them modifying it using [Visitor](https://github.com/webonyx/graphql-php/blob/master/src/Language/Visitor.php).

If I may suggest something, it would be interesting to have a callback for enter/leave like Visitor already in buildSchema.

**Comment by spawnia** [comment]
> If I may suggest something, it would be interesting to have a callback for enter/leave like Visitor already in buildSchema.

Can you elaborate? What new use cases would this enable/how does it make existing use cases easier?

**Comment by brunobg** [comment]
Of course! The way I implemented the class loader I described above was more or less like this (and I am still struggling to find out how to replace the Type class on the fields):

```php
$ast = \GraphQL\Language\Parser::parse($data);
Visitor::visit($this->ast, [
    NodeKind::SCALAR_TYPE_DEFINITION => function ($node) {
        // some code to read the directives and map the $node->name->value to my class
       return null;
    }
]);
$schemaBuilder = new \GraphQL\Utils\BuildSchema(ast);
$schema = $schemaBuilder->buildSchema();
```

A visitor is what I expected the callback from `BuildSchema` to do -- it's not clear to me from the docs what exactly it does, is it only  called for types?

Again, the use cases is  what is being done in the tests. You build a schema manually in the VariablesTest with the custom class:

https://github.com/webonyx/graphql-php/blob/1f0ec2408010e01879b1f980afac65de2344c631/tests/Executor/VariablesTest.php#L132

but from what I understood from your answer it's hard to do that from a graphql description. Lighthouse seems to reimplement the schema builder completely. If there was a Visitor-like callback for BuildSchema, I could have a callback that could easily customize the creation of the `Schema` object, replacing the generic Scalar type for a MyCustomClass().

Thank you for the interest and prompt replies.

**Comment by vladar** [comment]
I might have understood this incorrectly. But my understanding is that:

1. You are trying to define your schema with SDL
2. Define a class for your custom scalar type using a directive

Technically this should be probably done by allowing custom `typeLoader` for the `buildSchema`. But in the meantime I think there is an existing workaround - you can replace the `typeLoader` and replace `CustomScalarType` with your custom class.

Something along the lines should probably work (didn't check though):

```php
$ast = \GraphQL\Language\Parser::parse($sdl);
$schemaBuilder = new \GraphQL\Utils\BuildSchema($ast);
$schema = $schemaBuilder->buildSchema();

$originalTypeLoader = $schema->getConfig()->typeLoader;

$schema->getConfig()->typeLoader = function($typeName) use ($originalTypeLoader) {
    $type = $originalTypeLoader($typeName);
    if ($type instanceof CustomScalarType) {
        $directives = $type->astNode->directives;
        // Look for your `@scalar` directive in $directives and return your custom class instead
        $customClass = findMyScalarClassName($directives);
        return new $customClass($type->config);
    }
    return $type;
}
```

Obviously just a workaround but it should work?

**Comment by brunobg** [comment]
Thank you very much, it seems that will work perfectly.

**Comment by brunobg** [comment]
Ahn, @vladar, sorry for the endless pestering, but besides the bug from #683, it seems that this does not replace types in fields. For example:

```graphql
scalar TestText @scalar(class: "ModelariumTests\\Laravel\\ScalarTestText")

type User {
  id: ID!
  name: String!
  description: TestText!
}
```

calls your `typeLoader` only with a `ObjectType`, and when I traverse the AST the `FieldDefinition` `$type` is still `CustomScalarType`.

Perhaps I'm missing something and it only lazy loads? 

Could I use a `SchemaExtender` to do this, overriding `extendCustomScalarType`?

**Comment by vladar** [comment]
Good catch. `typeLoader` is only partially useful indeed.

Reopening since I think this is a typical use-case (providing custom scalar type class) so we should provide a way to do it in this library.

**Comment by spawnia** [comment]
I am attempting to solve this with https://github.com/webonyx/graphql-php/pull/1872, looking for feedback.
